### DISCUSSION TITLE
Some observations on the operation of the display on the ILI9488 and ILI9341 controller and esp32-s3

### DISCUSSION THREAD

**Discussion by brightproject** [Bodmer/TFT_eSPI#3439] [Q&A - General]
Recently I conducted several experiments with the `ESP32-S3-N16-R8` microcontroller and displays based on the `ILI9488` and `ILI9341`.

![ili9341_ili9488](https://github.com/user-attachments/assets/9c914ca1-623f-4709-82cf-e10c7dc4ff44)

I designed a printed circuit board according to the following settings:

> #define ESP32_PARALLEL
> // #define TFT_PARALLEL_8_BIT
> 
> #define DISABLE_ALL_LIBRARY_WARNINGS
> 
> // #define ILI9341_DRIVER // DB8 = TFT_D0, DB9 = TFT_D1, DB10 =TFT_D2, DB11 =TFT_D3, DB12 =TFT_D4, DB13 = DB5, DB14 = TFT_D6, DB15 = TFT_D7, IM2 = 0, IM1 = 0, IM0 = 1
> #define ILI9488_DRIVER // DB0 = TFT_D0, DB1 = TFT_D1, DB2 =TFT_D2, DB3 =TFT_D3, DB4 =TFT_D4, DB5 = DB5, DB6 = TFT_D6, DB7 = TFT_D7, IM2 = 0, IM1 = 1, IM0 = 1
> 
> #define TFT_CS    5
> #define TFT_DC    7
> 
> #define TFT_RST  -1// in scheme +3.3V
> 
> #define TFT_WR    4
> #define TFT_RD    10
> 
> #define TFT_D0   14
> #define TFT_D1   13
>  #define TFT_D2   47
> #define TFT_D3   38
> #define TFT_D4   39
> #define TFT_D5   40
> #define TFT_D6   41
> #define TFT_D7   42

The code compiles with warnings:
```
.pio/libdeps/esp32-s3-devkitc1-n16r8/TFT_eSPI/Processors/TFT_eSPI_ESP32_S3.h:398:34: note: in expansion of macro 'GPIO_DIR_MASK'
       #define GPIO_OUT_CLR_MASK (GPIO_DIR_MASK | (1 << TFT_WR))
```

Accordingly, there is no image on the display.
I had to re-solder the connection diagram, according to the following:
> #define ESP32_PARALLEL
> // #define TFT_PARALLEL_8_BIT
> 
> #define DISABLE_ALL_LIBRARY_WARNINGS
> 
> // #define ILI9341_DRIVER // DB8 = TFT_D0, DB9 = TFT_D1, DB10 =TFT_D2, DB11 =TFT_D3, DB12 =TFT_D4, DB13 = DB5, DB14 = TFT_D6, DB15 = TFT_D7, IM2 = 0, IM1 = 0, IM0 = 1
> #define ILI9488_DRIVER // DB0 = TFT_D0, DB1 = TFT_D1, DB2 =TFT_D2, DB3 =TFT_D3, DB4 =TFT_D4, DB5 = DB5, DB6 = TFT_D6, DB7 = TFT_D7, IM2 = 0, IM1 = 1, IM0 = 1
> 
> #define TFT_CS    5
> #define TFT_DC    7
> 
> #define TFT_RST  -1// in scheme +3.3V
> 
> #define TFT_WR    4
> #define TFT_RD    10
> 
> #define TFT_D0   14
> #define TFT_D1   13
> #define TFT_D4   16
> #define TFT_D5   3
> #define TFT_D6    9
> #define TFT_D7    8

It turns out that the pins

`47, 38, 39, 40, 41, 42`

Cannot be reassigned by the mask for display output.

> https://www.espressif.com/sites/default/files/documentation/esp32-s3-wroom-1_wroom-1u_datasheet_en.pdf

![ESP32-S3-WROOM-1_3_pins](https://github.com/user-attachments/assets/8554488c-c71c-44fa-9bca-34071ae2177e)
![ESP32-S3-WROOM-1_3](https://github.com/user-attachments/assets/986fe191-c668-4e10-8120-96435ce66990)
I also did a little experiment connecting the `ILI9341` display via `SPI.`
The platformio settings are as follows:

> [env:esp32-s3-devkitc1-n16r8]
> ; platform = espressif32 @ 5.3.0
> platform = espressif32 @ 6.6.0
> board = esp32-s3-devkitc1-n16r8
> framework = arduino
> monitor_speed = 115200
> monitor_port = COM6
> upload_speed = 921600
> upload_port = COM22
> monitor_filters = esp32_exception_decoder
> 
> lib_deps = bodmer/TFT_e SPI@^2.5.43

If version `5.3.0` 
`Processing esp32-s3-devkitc1-n16r8 (platform: espressif32 @ 5.3.0; board: esp32-s3-devkitc1-n16r8; framework: arduino)` 
then the output [example](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/480%20x%20320/UTFT_Demo_480x320/UTFT_Demo_480x320.ino) with a display connected via `SPI` is `1363 msec`.
And if you update to version `6.6.0`
`Processing esp32-s3-devkitc1-n16r8 (platform: espressif32 @ 6.6.0; board: esp32-s3-devkitc1-n16r8; framework: arduino)`
then the output will be `842 msec`.

The slowdown of the example occurs at the last step, where the pixels are output

```
  tft.fillRect(1,15,478-1,304-15,TFT_BLACK);

  for (int i=0; i<10000; i++)
  {
    tft.drawPixel(2+random(476), 16+random(289),random(0xFFFF));
  }
```

It seems that the `8-bit parallel` interface should be faster than the `SPI`, but on the `480x320` display I see something different.

For display `ILI9341` 320х240, output via 8-bit parallel interface occurs in `542 ms`.
There are no delays.
For display `IILI9488` 480х320, output via 8-bit parallel interface occurs in `1510 ms`.
There is a slowdown again on the pixel output in the very last test.
![ili9488_8_bit_parallel](https://github.com/user-attachments/assets/b82861a1-9ae2-4650-8299-1ba509070bc4)

https://github.com/user-attachments/assets/5f2dcf91-421e-4ef4-b653-95b27a2a7ee7

I just decided to write about my observations, maybe it will help someone or they will give me some tips on how to increase the speed of the displays.

**Comment by dkpetrutsik-boop** [comment]
Hello! What libraries did you use to connect the ILI9488? Is there a test sketch to check the connection between the ESP32-S3 and the ILI9488 display? I wanted to make a speedometer+tachometer+speed measurement. It worked with the ESP32, but it was slow (due to limited memory and the absence of PSRAM). Therefore, I purchased the ESP32-S3 N16R8, but I am unable to solve the issue of connecting the display and displaying anything on it.

**Reply by brightproject** [reply]
> What libraries did you use to connect the ILI9488?

https://github.com/Bodmer/TFT_eSPI/blob/master/examples/480%20x%20320/UTFT_Demo_480x320/UTFT_Demo_480x320.ino
